NineAnimator/Models/masterani.me/SearchMasterAnime.swift

```diff
@@ -0,0 +1,95 @@
+//
+//  This file is part of the NineAnimator project.
+//
+//  Copyright © 2018 Marcus Zhou. All rights reserved.
+//
+//  NineAnimator is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  NineAnimator is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with NineAnimator.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+import Foundation
+
+class NASearchMasterAnime: SearchProtocol {
+    static let apiPathSearch = "/api/anime/filter?search=%@&order=score_desc&page=%@"
+    
+    var query: String
+    
+    var totalPages: Int?
+    
+    var availablePages: Int { return _results.count }
+    
+    var moreAvailable: Bool { return totalPages == nil || _results.count < totalPages! }
+    
+    var delegate: SearchPageDelegate?
+    
+    private let _parent: NASourceMasterAnime
+    
+    private var _lastRequest: NineAnimatorAsyncTask?
+    
+    private var _results = [[AnimeLink]]()
+    
+    init(query: String, parent: NASourceMasterAnime) {
+        self.query = query
+        self._parent = parent
+        more()
+    }
+    
+    deinit{ _lastRequest?.cancel() }
+    
+    func animes(on page: Int) -> [AnimeLink] {
+        return _results[page]
+    }
+    
+    func more() {
+        if _lastRequest == nil && moreAvailable {
+            let keyword = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)!
+            let path = String(format: NASearchMasterAnime.apiPathSearch, keyword, "\(availablePages + 1)")
+            _lastRequest = _parent.request(ajax: path){
+                response, error in
+//                [weak self] response, error in
```

Comment:
Do we still need to make sure no strong reference cycle here?

Reply:
Yeah, I guess. The debugger doesn't work well with weak references.

Reply:
That sounds weird. Might need to spend more time on that then.